### PR TITLE
Bug 1939054: Disable startup timeout for Spot MHC

### PR DESCRIPTION
**Pull request by JoelSpeed**
The spot termination handler MHC should only ever remove Machines that meet the spot termination condition. We do not want it to remove machines because they took too long to start up. That should be a user decision.

This PR adds the ability to disable the node startup timeout completely, and uses that for the spot termination MHC.

### CODE REVIEW

**Comment by JoelSpeed** [comment]
/bugzilla refresh

CC @n1r1 I think you might be interested in this

File: install/0000_30_machine-api-operator_07_machinehealthcheck.crd.yaml [hunk 1]
```diff
@@ -104,7 +104,7 @@ spec:
                       type: string
                     timeout:
                       description: Expects an unsigned duration string of decimal numbers each with optional fraction and a unit suffix, eg "300ms", "1.5h" or "2h45m". Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
```
Comment (lobziik) [inline install/0000_30_machine-api-operator_07_machinehealthcheck.crd.yaml:106]: Suggest to extend this description and describe "-1" behaviour there.

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller.go [hunk 2]
```diff
@@ -629,12 +629,17 @@ func (t *target) needsRemediation(timeoutForMachineToHaveNode time.Duration) (bo
 		if t.Machine.Status.LastUpdated == nil {
 			return false, timeoutForMachineToHaveNode, nil
 		}
-		if t.Machine.Status.LastUpdated.Add(timeoutForMachineToHaveNode).Before(now) {
+		if timeoutForMachineToHaveNode != time.Duration(-1) && t.Machine.Status.LastUpdated.Add(timeoutForMachineToHaveNode).Before(now) {
+			fmt.Printf("Timeout: %v", timeoutForMachineToHaveNode)
```
Comment (lobziik) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller.go:633]: Debug print?

File: install/0000_30_machine-api-operator_07_machinehealthcheck.crd.yaml (hunk 1 above)
Comment (JoelSpeed) [reply install/0000_30_machine-api-operator_07_machinehealthcheck.crd.yaml:106]: I'd rather keep this undocumented for now, let's test it out with just our MHC and then negotiate with upstream to get feature parity, once we have done that, we can document it

**Review by elmiko — APPROVED** [review]
this makes sense to me, and while i generally agree with @lobziik about the documenting of this feature i think we can make an exception given @JoelSpeed's reasoning in this case.
/approve

**Comment by elmiko** [comment]
these test failures seem unrelated to this change, but i'm not 100% sure yet.

**Comment by elmiko** [comment]
digging in to the must-gather on one of the failures, i see this
```
Error creating: pods "aws-ebs-csi-driver-node-" is forbidden: unable to validate against any security context constraint: [provider "anyuid": Forbidden: not usable by user or serviceaccount, provider restricted: .spec.securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used, spec.volumes[0]: Invalid value: "hostPath": hostPath volumes are not allowed to be used, spec.volumes[1]: Invalid value: "hostPath": hostPath volumes are not allowed to be used, spec.volumes[2]: Invalid value: "hostPath": hostPath volumes are not allowed to be used, spec.volumes[3]: Invalid value: "hostPath": hostPath volumes are not allowed to be used, spec.containers[0].securityContext.privileged: Invalid value: true: Privileged containers are not allowed, spec.containers[0].securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used, spec.containers[0].securityContext.containers[0].hostPort: Invalid value: 10300: Host ports are not allowed to be used, spec.containers[1].securityContext.privileged: Invalid value: true: Privileged containers are not allowed, spec.containers[1].securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used, spec.containers[1].securityContext.containers[0].hostPort: Invalid value: 10300: Host ports are not allowed to be used, spec.containers[2].securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used, spec.containers[2].securityContext.containers[0].hostPort: Invalid value: 10300: Host ports are not allowed to be used, provider "nonroot": Forbidden: not usable by user or serviceaccount, provider "hostmount-anyuid": Forbidden: not usable by user or serviceaccount, provider "machine-api-termination-handler": Forbidden: not usable by user or serviceaccount, provider "hostnetwork": Forbidden: not usable by user or serviceaccount, provider "hostaccess": Forbidden: not usable by user or serviceaccount, provider "node-exporter": Forbidden: not usable by user or serviceaccount, provider "privileged": Forbidden: not usable by user or serviceaccount]
```
not exactly sure what this means yet.

**Comment by elmiko** [comment]
seems like we are hitting this bug, https://bugzilla.redhat.com/show_bug.cgi?id=1913069

**Comment by JoelSpeed** [comment]
/hold There is something wrong with the regex here, this needs further investigation
```
MachineHealthCheck.machine.openshift.io "machine-api-termination-handler" is invalid: spec.nodeStartupTimeout: Invalid value: "-1": spec.nodeStartupTimeout in body should match '^([0-9]+(\.[0-9]+)?(ns|us|µs|ms|s|m|h))+$'
```
The regex doesn't match the regex that is in the CRD manifest in the release image, so not sure why this error is happening in the CVO logs

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller.go [hunk 3]
```diff
@@ -738,6 +748,12 @@ func (t *target) needsRemediation(timeoutForMachineToHaveNode time.Duration) (bo
 
 	// the node has not been set yet
 	if t.Node == nil {
+		if timeoutForMachineToHaveNode.Seconds() == disabledNodeStartupTimeout.Seconds() {
+			// Startup timeout is disabled so no need to go any further.
+			// No node yet to check conditions, can return early here.
```
Comment (lobziik) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller.go:753]: Might it worth to have some log message here?

**Comment by JoelSpeed** [comment]
/hold cancel

We have bene working with the upstream and I think we've agreed on this disable via zero value feature

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller.go (hunk 3 above)
Comment (JoelSpeed) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller.go:753]: In the rest of this logic, we only log when the machine is actually being remediated. When we determine it's still healthy of possibly needs a check later, we don't log. 

I'd be tempted here to keep this consistent here

**Pull request merged**